### PR TITLE
Add day  14 blog

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**

### CODE REVIEW

File: docs/2023/puzzles/day14.md [hunk 1]
```diff
@@ -2,10 +2,357 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+    val array = Array.fill(extentX)(Array.fill(extentY)('.'))
+    for
+      (line, y) <- lines.zipWithIndex
+      (char, x) <- line.zipWithIndex
+    do array(x)(y) = char
+    dataN = array
+
+  override def toString =
+    val sb = new StringBuilder
+    for y <- (0 until extentY).toIterator do
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:63]: doesn't compile

File: docs/2023/puzzles/day14.md [hunk 2]
```diff
@@ -2,10 +2,357 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+  def setState(newState: String): Unit =
+    val lines = newState.split('\n').toList
+    val extentX = lines(0).length
+    val extentY = lines.length
+    val array = Array.fill(extentX)(Array.fill(extentY)('.'))
+    for
+      (line, y) <- lines.zipWithIndex
+      (char, x) <- line.zipWithIndex
+    do array(x)(y) = char
+    dataN = array
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:59]:
```suggestion
  def setState(newState: String): Unit =
    dataN = newState.split('\n').map(_.toArray)
```
?

File: docs/2023/puzzles/day14.md [hunk 3]
```diff
@@ -2,10 +2,357 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+
+  override def toString =
+    val sb = new StringBuilder
+    for y <- (0 until extentY).toIterator do
+      for x <- (0 until extentX).toIterator do
+        sb.append(this(x, y))
+      sb.append('\n')
+    sb.toString
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:67]:
```suggestion
    dataN.map(_.mkString).mkString("\n")
```
?

File: docs/2023/puzzles/day14.md [hunk 4]
```diff
@@ -2,10 +2,357 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+```scala
+def cycle(model: Model, times: Int): Unit =
+  val cache: collection.mutable.Map[String, String] = collection.mutable.Map.empty
+  var currentState = model.toString
+  for i <- 1 to times do
+    if i % 10_000_000 == 0 then
+      println(s"Cycle $i of ${times}; cache size: ${cache.size}")
+
+    if cache.keySet.contains(currentState) then
+      currentState = cache(currentState)
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:209]: actually at the first time you get a cache hit you can compute the result at 1 billion in 1 step and return

Comment (bishabosha) [reply docs/2023/puzzles/day14.md:209]: see https://github.com/bishabosha/advent-of-code-2023/blob/a612e0dc72943e95677ab924243677ebafc2b12e/2023-day14.scala#L93C1-L108C10

**Comment by bishabosha** [comment]
btw this solution does the entire rotation and cycling in about 2 lines:
https://github.com/merlinorg/aoc2023/blob/main/src/main/scala/Day14.scala

File: docs/2023/puzzles/day14.md (hunk 1 above)
Comment (anatoliykmetyuk) [reply docs/2023/puzzles/day14.md:63]: Strange, works fine on my machine. Are you using the Scala 3.3.1? What error are you getting?

File: docs/2023/puzzles/day14.md (hunk 2 above)
Comment (anatoliykmetyuk) [reply docs/2023/puzzles/day14.md:59]: What my code does is ensuring the `array(x)(y)` syntax. That is, when addressing the array, columns get addressed first (`x`), and rows - second (`y`). This is more in line with the mathematical view of the problem as a coordinate system, where the convention is to address the horizontal axis first. The simpler `newState.split('\n').map(_.toArray)` would have rows - the vertical axis - addressed first.

Going into a bit of trouble getting the coordinates right allows to use `x` and `y` in code, as opposed to `row` and `col`, which helps with understanding.

File: docs/2023/puzzles/day14.md (hunk 3 above)
Comment (anatoliykmetyuk) [reply docs/2023/puzzles/day14.md:67]: `toString` here returns the *rotated* representation of the platform (see `this(x, y)`), as specified by the `rotation` variable`. `dataN.map(_.mkString).mkString("\n")` is not equivalent, as it always returns the default rotation of the platform - facing North.

File: docs/2023/puzzles/day14.md (hunk 4 above)
Comment (anatoliykmetyuk) [reply docs/2023/puzzles/day14.md]: Nice! I implemented the shortcut solution.

File: docs/2023/puzzles/day14.md (hunk 1 above)
Comment (bishabosha) [reply docs/2023/puzzles/day14.md:63]: this snippet has no field called extendY or extentX

File: docs/2023/puzzles/day14.md (hunk 2 above)
Comment (bishabosha) [reply docs/2023/puzzles/day14.md:59]: right, sorry I didn't notice the y and x swap

Comment (bishabosha) [reply docs/2023/puzzles/day14.md:59]: but you will get the same behaviour with a transpose I believe:
```scala
scala> "012\n345".split('\n').map(_.toArray).transpose
val res3: Array[Array[Char]] = Array(Array(0, 3), Array(1, 4), Array(2, 5))
```

Comment (anatoliykmetyuk) [reply docs/2023/puzzles/day14.md]: Right, transpose also worked. I updated the article.

File: docs/2023/puzzles/day14.md [hunk 5]
```diff
@@ -2,10 +2,337 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+The approach from Part 1 doesn't work here: we now need to modify our model state. The reason we didn't have to do it in Part 1 is that our model evolves in only a single step, and we only need one metric from the end state, so we don't have to calculate the entire final state. In Part 2, however, the model evolves in many steps, and each successive step depends on the previous step. So, we need to calculate the entire final state.
+
+### The Model
+Actually, we already know how to do the Northern tilt - similarly to Part 1. We do not want to re-implement the tilts for the other directions - instead, we want to use the Northern tilt as a building block for the other tilts. So, we need to be able to rotate our model in all four directions, and then tilt it North.
+
+We can represent the model as a 2D array of chars:
+
+```scala
+class Model(state: String):
+  private var dataN: Array[Array[Char]] = Array.empty
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:47]: maybe add a comment that the array is indexed as `dataN(x)(y)` for a given point (x, y)

Comment (anatoliykmetyuk) [reply docs/2023/puzzles/day14.md]: Done

File: docs/2023/puzzles/day14.md [hunk 6]
```diff
@@ -2,10 +2,337 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+
+So, there's a possibility that the desired state will be achieved earlier than the billion cycles. Looks like a good case for a dynamic programming approach: we need to remember the states of the model we've seen before and what they look like after one cycle. And if the current state is in our cache, no need to compute the next state again.
+
+Furthermore, since the transitions between states are deterministic (a state A always leads to state B), the moment one state in our cache is followed by another, previously encountered, state from that cache, we can stop cycling and just calculate the final state from the cache.
+
+```scala
+import scala.util.boundary, boundary.break
+
+def cycle(model: Model, times: Int): Unit =
+  val chain = collection.mutable.ListBuffer.empty[String]
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:198]: If the aim here is all about optimisation shouldn't we not have a linear lookup?

File: docs/2023/puzzles/day14.md [hunk 7]
```diff
@@ -2,10 +2,337 @@ import Solver from "../../../../../website/src/components/Solver.js"
...
+  val model = Model(input)
+  rollUp(model)
+  val loading = totalLoading(model)
+  println(loading)
+
+def part2(input: String) =
+  val model = Model(input)
+  cycle(model, 1_000_000_000)
+  val loading = totalLoading(model)
+  println(loading)
```
Comment (bishabosha) [inline docs/2023/puzzles/day14.md:333]:
```suggestion
def part1(input: String): Int =
  val model = Model(input)
  rollUp(model)
  totalLoading(model)

def part2(input: String): Int =
  val model = Model(input)
  cycle(model, 1_000_000_000)
  totalLoading(model)
```

**Review by bishabosha — APPROVED** [review]
alright, looks nice!

**Pull request merged**